Simulate live-tracking of bus location

Dewar0019 opened the pull request:
Adds bus location data every 5 seconds to make it appear bus is driving around

-I'll merge this manually following [https://github.com/Vermonster/schoolbot/blob/master/CONTRIBUTING.md#pull-requests](url)

jpowers commented on config/initializers/bus_location_simulator.rb:
```diff
@@ -0,0 +1,38 @@
+    iteration += 1
+    iteration = 0 if iteration >= 360 # reset circle
+  end
+end
```
Nice! Can we move this implementation to it's own class? 
`class BusLocationSimulator` 

Then we can use the initializer to call it.

grantovich commented:
Looks like this could be quite useful for local testing! It always was a bit silly that you had to keep re-running the dev prime task as your test data dropped off the end of the 5-minute window.

@Dewar0019 FYI, the `CONTRIBUTING` is a bit outdated now, since GitHub added a "Rebase and merge" option to the merge pulldown (this can be made the default for this repository, but I don't have the permissions to do it myself).

We seem to have some things going wrong on CI, any idea what that's about? Unfortunately I'm not in a good position to look into it until next week.

Dewar0019 commented:
Two of the feature tests are failing related to user map views. Unfortunately, it's only occurring  circleci and I haven't found a good way to debug it.

grantovich commented:
Hm. It seems there is other strange stuff going on, like the most recent build running out of memory. I'm willing to put this down to bit-rot and look into it later; we do have some fiddly customizations on Circle that have spontaneously broken in the past. Let's just ensure a full `rake` completes successfully in dev as the last thing before merging this.

bkaney commented:
Hey @Dewar0019 -- I am curious why you chose to use rufus-scheduler instead of whenever (https://github.com/javan/whenever)?

Dewar0019 commented:
I basically went off this blog post
http://atlwendy.ghost.io/set-up-a-cron-job-in-rails/
Someone mentioned that 'whenever; reloads the environment every time it executes which might be bad? And Rufus-scheduler had an option to halt the process if needed which I didn't find in the docs for whenever.

Dewar0019 commented:
@grantovich I believe it has to do something related to QT5 and Qmake. Previously I had to modify some of the pre-dependencies in circle.yml just to get the tests to start running.

grantovich commented:
@Dewar0019 Let's keep this PR focused and only include the live-tracking-simulation changes. CI is an unrelated problem that should be the subject of a separate PR (and which I'm willing to look into myself, once I'm back in the country next week). If this PR isn't particularly urgent, we could wait to complete the CI fixes, then rebase these changes.

@bkaney I'm lacking context on this change; is the simulator intended to be used outside the dev environment, anyway? I don't think it would work as expected if multiple server processes were running.

grantovich commented:
It also occurs to me that we're already using Clockwork for scheduled tasks, and that would be preferable in the event that we wanted to use this outside of dev (which I'm still unclear on). For use in dev, we'd want to re-examine the consequences of running the clock process as part of `bin/serve`, which we currently don't.

Dewar0019 commented:
@grantovich I made the switch over to Clockwork and added the clock process into bin/serve for dev. I also added a conditional to exclude the Zonar updateBusLocation when in development. Let me know if this is okay and if there are other things I should tweak.

Dewar0019 commented:
I had originally only planned to use this for development but there was some mention of using it in staging for demo purposes.

grantovich commented:
Gotcha. Right now the staging environment closely matches production, in that the seed data assumes schools have been loaded and Zonar is providing live bus locations – so we might need some further tweaking to make the simulator usable there.

For now, what if we just set things up so the `ENV` value switches the simulator on, and *also* switches the Zonar updates off? That will at least ensure they're not fighting each other.

pdw207 commented:
@Dewar0019  I think Alex's last comment makes sense. Can you make the final change and merge?

grantovich commented:
Just curious, are others not seeing the open review comments I've made on the diff? For me they show up grouped together earlier on the timeline as a "pending review". This is the first time we've had a PR when this new review format was in place, so I'm not sure how it appears for others.

pdw207 commented:
@grantovich I don't see your comments.

grantovich commented:
@pdw207 Thanks, I checked into it a bit more and it turns out I was doing it wrong – if you submit diff comments using the "Start review" button, it queues them up in a state where only you can see them, and then you have to hit "Submit review" to let other people see them. So now those comments I made, like, weeks ago are finally visible 😅 Sorry for the confusion!

Dewar0019 commented on .gitignore:
```diff
@@ -10,3 +10,4 @@
 /public/system
 /tags
 /tmp/*
+*.rdb
```
Yes, I believe so. According to the Redis documents it is a snapshot is produced when n number of records change in x time. https://redis.io/topics/persistence#snapshotting

grantovich replied:
Interesting. And this was being dumped in the app directory? I wonder why that started showing up on this branch. No harm in adding it, I suppose.

Dewar0019 commented on lib/clock.rb:
```diff
@@ -6,16 +6,28 @@ module Clockwork
+    iteration = 0
+    every(5.seconds, 'SimulateBusLocation') do
+      Bus.find_each do |bus|
+        SimulateBusLocationJob.perform_now(bus, iteration)
```
👍

Dewar0019 commented on lib/clock.rb:
```diff
@@ -6,16 +6,28 @@ module Clockwork
 
+  if ENABLED_BUS_SIMULATOR
+    iteration = 0
+    every(5.seconds, 'SimulateBusLocation') do
```
👍

Dewar0019 commented on lib/clock.rb:
```diff
@@ -6,16 +6,28 @@ module Clockwork
     BusLocation.where('created_at < ?', 1.week.ago).delete_all
   end
 
+  if ENABLED_BUS_SIMULATOR
```
👍

Dewar0019 commented on lib/clock.rb:
```diff
@@ -6,16 +6,28 @@ module Clockwork
+  if ENABLED_BUS_SIMULATOR
+    iteration = 0
+    every(5.seconds, 'SimulateBusLocation') do
+      Bus.find_each do |bus|
```
👍

Dewar0019 commented on .rubocop.yml:
```diff
@@ -7,6 +7,7 @@ AllCops:
   - "vendor/**/*"
   - "db/**/*"
   - "client/**/*"
+  - "app/jobs/simulate_bus_location_job.rb"
```
The calculation for the 'new_latitude' and 'new_longitude' kept triggering an error "Assignment Branch Condition size for perform is too high". I tried refactoring but it was becoming more and more complicated.

grantovich replied:
Oh, yeah, I had meant to rework things so the complexity metrics wouldn't actually break the build on Circle, since we do have CodeClimate. In the meantime if a metrics cop is giving you trouble you can [disable it with a magic comment](http://rubocop.readthedocs.io/en/latest/configuration/#disabling-cops-within-source-code) (here's the [list of cop names](http://rubocop.readthedocs.io/en/latest/cops_metrics/)). That way you still get the benefit of the style checking.

grantovich commented:
If we could address the Rubocop issue in a more targeted manner [as I described here](https://github.com/Vermonster/schoolbot/pull/111#discussion_r104537326) that would be nice, but this is looking good. As usual let's rebase into a small number of meaningful commits (probably just one, in this case, hard to break it down further) before merging.
